### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by stebenne** [guyfrancoeur/INF3135_H2019_TP2#2]
Bonjour à tous,
Comme discuter dans le cours, nous avons vue comment améliorer notre algorithme pour trouver des nombres parfaits. 

Par contre, ce n'est pas suffisant pour être en mesure de calculer 10000 à 140000000000 en 90 secondes.

La seule solution que je vois, serait d'utiliser la bibliothèque standard **<threads.h>** pour faire du _multithreading_. Comme ce n'est pas quelque chose que nous avons vue en classe, je ne vois pas comment réussir le test data2.in en 90 secondes.

Pour le moment, je suis en mesure de faire 1 à 10000000 en 1min 20sec.

**Comment by stebenne** [comment]
Question de : **LauriGui**

> Bonjour,

Je ne sais pas comment formuler ma question, sans demander de réponse, mais je n'arrive vraiment pas à accélérer suffisamment mon programme, pour le test 2... J'ai même essayé de seulement afficher tous les nombres, compris dans l'intervalle donnée, dans une boucle **for** super simple, mais juste ça, ça prend trop de temps... En faisant des recherches, certains disaient qu'un simple affichage ralentit le processus, mais peu importe que j'affiche ou pas, ça me donne le même temps... Donc, je ne vois pas comment faire autrement, sans faire de détour, du même genre qu'ignorer tous les nombres impairs, et sans rendre ça trop complexe...

Merci!

**Comment by charimene** [comment]
Bonjour a tous,
moi aussi j'ai les memes difficultés, en cherchant sur internet j'ai trouve une formule mathématique en fonction de nombres premiers qui permet de trouver beaucoup plus rapidement les nombres parfaits qu'on cherche, mais je sais pas si on a le droit de l'utiliser ...
Merci :)

**Comment by stebenne** [comment]
@charimene , Je crois que d'utiliser une formule pour détecter si c'est un nombre premier va ralentir d'avantage l'execution du programme (Des formules que j'ai trouvé, ça semble compliquer à coder).

Voici mon idée, (Je vais essayer d'être un peu vague, question de pas donner une réponse direct.)
Prenons **1187** qui est un nombre premier.

Si tu utilise la logique donnée en classe, soit **Racine de 1187 = 34** qui te donne le plus grand diviseur 
entier à tester.

Il à moyen de déduire que le nombre que tu traites, est soit premier ou simplement non parfait avant d'atteindre le dernier diviseur.

Donc, disons qu'après 20 tours de boucle à tester des diviseurs, tu as encore une **somme des diviseurs** égal à ton point de départ. Il est fort probable que ton nombre est soit premier ou simplement pas parfait.

Reste que ça ne règle pas le problème de faire des milliard d'opérations en 90sec, mais ça peux aider à accélérer le tout.

**Comment by robinstpierre** [comment]
J'ai aussi trouvé la méthode des nombres premier, par contre cette méthode ne permet de trouver que des nombre parfait **pairs**. Et comme spécifier dans la FAQ du TP1, je crois qu'on ne peut prendre pour aquis qu'il n'existe pas de nombre parfait impairs.

**Comment by charimene** [comment]
@stebenne , @robinstpierre , merci pour vos réponses. :)

**Comment by guyfrancoeur** [comment]
@robinstpierre, les autres,

Il est important de différencier nombres parfaits pairs, et ne pas tester pour des nombres impairs.

Je crois que la science est claire sur la détection des nombres premiers.  Elle est aussi claire pour les nombres parfaits.  Il existe une relation qui est bien documentée et qui ne date de 2018.   Les mathématiciens ont sus prouvé un certain nombre de choses qui sont utilisables dans vos TPs.

Nous sommes des informaticiens en devenir.  La science et les mathématiques sont pour nous importantes.
Nous devons comprendre et utiliser les découvertes, la science et les maths pour faire des programmes qui seront à la fois performants dans la production de résultats, mais aussi dans l'écriture de ceux-ci.

Continuez vos réflexions, utilisez la science.

Soyez certain d'y aller par étapes.  L’optimisation de logiciel est un processus qui doit-ce faire lorsque les bases sont solides.

Basés vos réalisations sur des articles (scientifique) reconnus.

**Comment by LouisEtienneLemire** [comment]
Pour le moment, je comprends qu'un ''nombre premier est un diviseur d'un nombre parfait''. Et que lorsque l'on multiplie le ''diviseur nombre premier'' par le ''diviseur précédent'' ça donne un ''nombre parfait pair''. Mais comment intégrer la recherche des nombres parfaits impairs dans le calcul? C'est là la difficulté de rejoindre performance et rigueur à la fois...

**Comment by LouisEtienneLemire** [comment]
Tous les nombres premiers ne sont pas nécessairement des diviseurs de nombres parfaits? Je viens de faire un algorithme qui produit les nombres premiers... et je constate qu'il y en a bcp qui ne sont pas utilisé pour trouvé les premiers nombres parfaits... Toutefois il n'y a que le chiffre deux qui soit pair et premier... les autres chiffres premiers sont tous impairs! Pour trouver un nombre parfait impair on doit avoir deux nombres impairs...

**Comment by LouisEtienneLemire** [comment]
Si on peut vérifier la parité d'une séquence de diviseurs d'un nombre parfait en question et qu'ils sont pairs jusqu'à celui qui est premier et impairs, il suffirait de multiplier le nombre diviseur du nombre parfait précédent et le nombre diviseur premier impair pour obtenir le nombre parfait en question. Il faut donc trouver l'algorithme des multiples d'un nombre parfait. Je crois que c'est ce que Euler a démontré. Toutefois cette logique ne s'applique qu'aux nombres parfaits pairs...

**Comment by dbrsseau** [comment]
J'ai trouvé ce [document](https://www.univers-ti-nspire.fr/files/pdf/03-arith-parfaits-TNS21.pdf) qui explique la corrélation entre les nombres premiers et les nombres parfaits. (mathématiquement parlant) Il y a quelques formules intéressantes et bien très bien expliquées. Je crois qu'il est possible de retranscrire l'une de ses formules en code et de gagner considérablement en vitesse.

PS.: @guyfrancoeur y a-t-il un moyen de rajouter des informations au wiki (comme des liens utiles) ou si c'est réservé à ton utilisation seulement? Je ne sais pas pour les autres, mais moi je suis pas très fort en math. Je trouve ce liens pertinent pour ceux qui (comme moi) ont des lacunes en math.

**Comment by stebenne** [comment]
@dbrsseau , 
Merci beaucoup pour le partage de ta recherche.
Très bon document et très bien expliqué.

**Comment by fadifeghaly** [comment]
Marin Mersenne :  

> nombre parfait = 2^(p-1) * (2^p - 1)

**Comment by dbrsseau** [comment]
@fadifeghaly encore faut-il la comprendre... Sans le document cette formule ne veut rien dire...

**Comment by fadifeghaly** [comment]
Si premier de Mersenne --> t'appliques la formule pour trouver le nbr parfait! ça veut tout dire...

**Comment by dbrsseau** [comment]
@fadifeghaly je suis d'accord avec toi mais il ne faut pas donner de réponses ici... Aussi quelqu'un qui n'a pas lu le document ne sait aucunement le "p" fait référence à quoi...

**Comment by fadifeghaly** [comment]
C'est une simple indication, c'est loin d'être la réponse :)

**Comment by LouisEtienneLemire** [comment]
Dans la vérification des nombres premiers; en vérifiant les diviseurs, si il y en a plus de un le nombre est donc pas premier: en excluant 1 et lui-même. C'est un indice de condition d'optimisation de la recherche d'un nombre premier précis. Donc si tu trouves un diviseur autre que 1 et lui-même le nombre n'est pas premier. C'est à cet instant que tu sors. Peut-être trop évident... quand tu penses que 2 est un diviseurs de tous les nombres pairs. Ça élimine beaucoup de possibilité.

**Comment by albache** [comment]
Simplifier la boucle, c'est utile/bien, cependant ça m'évite de se casser la tête, mais c'est mauvais en performance car faire un algo en O(N) c'est quand même très différent !
Pour la realisation de ce tp, est-ce une obligation de passer par les nombres premiers?  

@guyfrancoeur

**Comment by dbrsseau** [comment]
Je sais pas si c'est une obligation mais la question tu dois te poser est : "respectes-tu le temps demandé de 90 d'exécution?" Si oui tant mieux, sinon tu devrais songer à l'algo! ;)
